### PR TITLE
[MiniMax-M3] Fuse SwiGLU and DeepGEMM MXFP8 quantization

### PR DESCRIPTION
**Pull request by Denji-kk**
## Motivation

MiniMax-M3's dense MLP and standalone shared experts materialize a BF16 SwiGLU activation before quantizing it for a DeepGEMM MXFP8 down-projection. Fuse those two operations to remove an intermediate activation write/read and one kernel launch.

## Modifications

- Add a Triton SwiGLU-OAI + group-32 MXFP8 kernel. Preserve alpha, clamp, the up-branch offset, and the BF16 rounding boundary; emit DeepGEMM's packed, MN-major, TMA-aligned UE8M0 scales.
- Extend the existing DeepGEMM linear wrapper to consume pre-quantized FP8 input with validated packed scales and BF16 output. Reject incompatible scales instead of sending them to the FlashInfer fallback.
- Enable the model fast path on SM100/SM103 for eligible DeepGEMM MXFP8 linears, contiguous 2D BF16 activations, and **1–1024 local tokens**. Other cases retain the original path, including larger batches, sequence parallel, and `torch.compile`.
- Keep the normal `down_proj` path for weight handling and collectives. The optimization covers dense MLPs and standalone shared experts; shared experts fused into routed MoE use a different path.

The switch is `SGLANG_OPT_MINIMAX_M3_FUSED_SWIGLU_MXFP8=0` to disable the optimization. It is separate from the existing ROCm switch. Eligible runs use `--fp8-gemm-backend=deep_gemm`; default `auto` backend selection is unchanged.

## Accuracy Tests

**62 GPU tests passed** on one B300, using the public [MiniMax-M3-MXFP8 configuration](https://huggingface.co/MiniMaxAI/MiniMax-M3-MXFP8/blob/c5454eb03678d8710e54a4e0fc681b9f3b4a3dba/config.json): H=6144, shared I=3072, dense I=12288.

Coverage includes exact FP8 values and packed scales; nondefault activation parameters; empty inputs; the actual MLP with TP1/TP4/TP8 local shapes; the 1024/1025-token boundary; CUDA Graph replay after an in-place weight-scale update; 3D, sequence-parallel, unquantized/SILU, and Dynamo (`backend="eager"`) fallback checks; and invalid pre-quantized layouts and bias handling.

All **36 benchmark cases** also produced bit-identical MLP outputs with the switch on/off (30 use fusion; six large-batch cases use the fallback). Inputs and weights are synthetic. Full-model task accuracy, distributed collectives, and serving throughput were not measured.

## Speed Tests and Profiling

Baseline: `c9fbe5f6552f26292f3ec3382a06cc1931f071cc` (GitHub main). Measurements use the same MLP with fusion off/on.

One NVIDIA B300 SXM6 AC; PyTorch 2.12.0a0+5aff3928d8.nv26.05 / CUDA 13.2, Triton 3.7.0, FlashInfer 0.6.18, sgl-deep-gemm 0.1.5.post3; TF32 enabled. These are the tested runtime versions, rather than an assertion that every dependency matches the main-branch pins.

CUDA-event timings: 20 eager warmups, graph capture, 200 ms replay warmup per variant, 11 randomized A/B/auto rounds, 10 replays per round, and 16 calls per graph for M<1024 or four otherwise. The table reports warm graph medians in microseconds. The benchmark writes all samples and eager timings to JSONL; eager timings include host launch gaps.

At M<=1024, 29 of 30 configurations reduce component latency by **2.96–15.57%**; shared TP1/M=1024 is effectively flat (+0.34%). `auto` selected FlashInfer CuTe DSL and is faster at many small-M points, so the benefit is specific to DeepGEMM configurations.

The 1024-token cap is intentional: a separate 21-round run of the unrestricted fusion at shared TP1/M=8192 measured 422.87 us unfused vs 445.96 us fused (-5.46%). Larger batches therefore retain the original implementation. Differences between separately captured baseline/PR fallback graphs below are timing variation, not fusion gains.

**TP is a local partition shape on one GPU. No all-reduce or routed/shared overlap is timed.** M is the number of local tokens; these are component latencies, not end-to-end serving numbers.

<details>
<summary>Full shape sweep (microseconds; positive reduction is faster)</summary>

| MLP | TP local shape | M | DeepGEMM baseline | This PR | Reduction | Default auto |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| shared | 1 | 1 | 21.62 | 20.52 | +5.09% | 15.95 |
| shared | 1 | 30 | 22.49 | 21.75 | +3.29% | 16.60 |
| shared | 1 | 240 | 26.48 | 25.45 | +3.86% | 25.78 |
| shared | 1 | 400 | 32.90 | 31.92 | +2.96% | 45.14 |
| shared | 1 | 1024 | 72.45 | 72.21 | +0.34% | 78.72 |
| shared | 1 | 8192 | 445.03 | 440.21 | fallback | 542.39 |
| shared | 4 | 1 | 18.30 | 17.51 | +4.30% | 13.72 |
| shared | 4 | 30 | 19.34 | 17.94 | +7.22% | 14.86 |
| shared | 4 | 240 | 21.41 | 19.93 | +6.88% | 18.96 |
| shared | 4 | 400 | 23.59 | 21.77 | +7.70% | 21.47 |
| shared | 4 | 1024 | 31.35 | 28.76 | +8.26% | 28.95 |
| shared | 4 | 8192 | 161.21 | 161.02 | fallback | 183.04 |
| shared | 8 | 1 | 17.66 | 16.85 | +4.56% | 13.60 |
| shared | 8 | 30 | 18.49 | 17.25 | +6.74% | 14.22 |
| shared | 8 | 240 | 20.51 | 19.28 | +5.98% | 18.06 |
| shared | 8 | 400 | 21.45 | 20.13 | +6.16% | 19.80 |
| shared | 8 | 1024 | 26.98 | 25.19 | +6.64% | 24.26 |
| shared | 8 | 8192 | 109.77 | 109.80 | fallback | 111.74 |
| dense | 1 | 1 | 52.13 | 50.49 | +3.14% | 48.23 |
| dense | 1 | 30 | 54.02 | 51.73 | +4.24% | 52.10 |
| dense | 1 | 240 | 90.06 | 80.97 | +10.10% | 94.45 |
| dense | 1 | 400 | 130.45 | 114.38 | +12.32% | 162.78 |
| dense | 1 | 1024 | 265.27 | 223.97 | +15.57% | 304.16 |
| dense | 1 | 8192 | 1908.62 | 1904.30 | fallback | 2340.80 |
| dense | 4 | 1 | 21.70 | 20.52 | +5.46% | 15.97 |
| dense | 4 | 30 | 22.53 | 21.33 | +5.33% | 16.86 |
| dense | 4 | 240 | 28.82 | 25.59 | +11.21% | 28.18 |
| dense | 4 | 400 | 36.16 | 31.36 | +13.28% | 46.75 |
| dense | 4 | 1024 | 78.55 | 69.59 | +11.41% | 85.67 |
| dense | 4 | 8192 | 496.15 | 496.34 | fallback | 603.62 |
| dense | 8 | 1 | 19.55 | 18.76 | +4.03% | 14.48 |
| dense | 8 | 30 | 20.36 | 19.08 | +6.27% | 15.64 |
| dense | 8 | 240 | 23.76 | 21.97 | +7.51% | 20.95 |
| dense | 8 | 400 | 26.32 | 23.85 | +9.37% | 25.55 |
| dense | 8 | 1024 | 42.01 | 36.65 | +12.76% | 48.40 |
| dense | 8 | 8192 | 271.76 | 271.90 | fallback | 319.39 |

</details>

### Reproduction

Run from this checkout on an SM100/SM103 GPU with SGLang's GPU dependencies installed:

```bash
PYTHONPATH=python SGLANG_JIT_DEEPGEMM_PRECOMPILE=0 \
  TORCHINDUCTOR_COMPILE_THREADS=4 \
  python test/registered/kernel/moe/test_minimax_m3_swiglu_deepgemm.py

PYTHONPATH=python SGLANG_JIT_DEEPGEMM_PRECOMPILE=0 \
  TORCHINDUCTOR_COMPILE_THREADS=4 \
  python benchmark/kernels/minimax_m3/bench_swiglu_mxfp8.py \
    --rows 1 30 240 400 1024 8192 --tp 1 4 8 --rounds 11 \
    --output /tmp/minimax-m3-swiglu-mxfp8.jsonl
```

## Checklist

- [x] Run pre-commit checks on all six changed files.
- [x] Add GPU correctness/regression tests and register them for `base-b-kernel-unit-test-4-gpu-b200`.
- [x] Document activation scope, fallback behavior, and reproduction commands.
- [x] Provide component accuracy and speed results, with measurement limitations stated above.
- [x] Follow SGLang code style guidance.

---
### CI States

Latest PR Test (Base): :x: [Run #34929230801](https://github.com/sgl-project/sglang/actions/runs/34929230801)
Latest PR Test (Extra): :x: [Run #34929230570](https://github.com/sgl-project/sglang/actions/runs/34929230570)
Latest PR Test (AMD ROCm 10): :x: [Run #34929231068](https://github.com/sgl-project/sglang/actions/runs/34929231068)

### CODE REVIEW

File: python/sglang/kernels/ops/moe/minimax_m3_swiglu_deepgemm.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+"""MiniMax M3 SwiGLU with BF16 rounding and DeepGEMM MXFP8 scales."""
+
+import torch
+import triton
+import triton.language as tl
+
+
+@triton.jit
+def _store_quant(
+    V, Q, S, ROW, COL, K: tl.constexpr, AM: tl.constexpr, BLOCK: tl.constexpr
```
Comment (zcnrex) [inline python/sglang/kernels/ops/moe/minimax_m3_swiglu_deepgemm.py:11]: Can we use lower case words for non-constexpr params?

**Review by zcnrex — COMMENTED** [review]
Could you remove the benchmark file and test file? Otherwise LGTM! Thanks.

File: python/sglang/kernels/ops/moe/minimax_m3_swiglu_deepgemm.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+def swiglu_quant(
+    x: torch.Tensor, alpha: float = 1.702, limit: float = 7.0
+) -> tuple[torch.Tensor, torch.Tensor]:
+    """Quantize non-interleaved SwiGLU to group-32 FP8 and packed MN-major scales."""
+    assert x.ndim == 2 and x.is_contiguous() and x.dtype == torch.bfloat16
+    assert x.is_cuda
+    m, two_k = x.shape
+    assert two_k > 0 and two_k % 256 == 0
+    k = two_k // 2
+    assert k % 128 == 0
```
Comment (zcnrex) [inline python/sglang/kernels/ops/moe/minimax_m3_swiglu_deepgemm.py:66]: Redundant assertion: two_k % 256 means k % 128
